File: cpp/src/tree.cpp
#include <algorithm>
#include <cstdlib>
#include <fmt/base.h>
#include <functional>
#include <ranges>

#include "sample.hpp"
#include "tree.hpp"

namespace fmm
{
FMMTree::FMMTree(const index_t depth, const double size)
{
  const auto& range = std::ranges::views::iota;

  data.clear();
  data.reserve(depth);

  for (size_t l = 0; l < depth; l++) {
    const index_t n = 1 << l;
    const auto fn = static_cast<double>(n);
    data.emplace_back(boost::extents[n][n][n]);

    for (index_t i = 0; i < n; i++) {
      for (index_t j = 0; j < n; j++) {
        for (index_t k = 0; k < n; k++) {
          data[l][i][j][k] = FMMCell(
            Vec3{ size * ((static_cast<double>(i) + 0.5) / fn - 0.5),
                  size * ((static_cast<double>(j) + 0.5) / fn - 0.5),
                  size * ((static_cast<double>(k) + 0.5) / fn - 0.5) },
            size / fn);
        }
      }
    }
  }

  for (index_t l = 0; l < depth; l++) {
    index_t width = 1 << l;
    for (index_t i = 0; i < width; i++) {
      for (index_t j = 0; j < width; j++) {
        for (index_t k = 0; k < width; k++) {
          FMMCell& cell = data[l][i][j][k];
          for (const index_t ni :
               range(std::max(i, 1l) - 1l, std::min(i + 2l, width)))
            for (const index_t nj :
                 range(std::max(j, 1l) - 1l, std::min(j + 2l, width)))
              for (const index_t nk :
                   range(std::max(k, 1l) - 1l, std::min(k + 2l, width)))
                cell.direct_neighbors.emplace_back(data[l][ni][nj][nk]);
          if (l > 0) {
            index_t pi = i / 2, pj = j / 2, pk = k / 2;
            for (const index_t ni :
                 range(2 * std::max(pi, 1l) - 2, std::min(width, 2 * pi + 4))) {
              for (const index_t nj : range(
                     2 * std::max(pj, 1l) - 2, std::min(width, 2 * pj + 4))) {
                for (const index_t nk : range(
                       2 * std::max(pk, 1l) - 2, std::min(width, 2 * pk + 4))) {
                  const FMMCell& near_child = data[l][ni][nj][nk];
                  if (
                    std::max({ std::max(ni, i) - std::min(ni, i),
                               std::max(nj, j) - std::min(nj, j),
                               std::max(nk, k) - std::min(nk, k) }) > 1)
                    cell.interaction_list.emplace_back(near_child);
                }
              }
            }
          }
        }
      }
    }
  }
}

index_t
FMMTree::depth() const
{
  return static_cast<index_t>(data.size());
}

index_t
FMMTree::tree_size() const
{
  return static_cast<index_t>(std::round(std::pow(8, depth()) - 1)) / 7;
}

double
FMMTree::space_size() const
{
  return data[0][0][0][0].size;
}

FMMCell&
FMMTree::get_leaf_from_pos(const Vec3& pos)
{
  index_t l = depth() - 1;
  index_t width = 1 << l;
  auto wf = static_cast<double>(width);
  double size = space_size();
  Vec3 local_pos = pos / size + Vec3{ 0.5, 0.5, 0.5 };
  index_t i = std::clamp(
    static_cast<index_t>(std::floor(X(local_pos) * wf)), 0l, width - 1);
  index_t j = std::clamp(
    static_cast<index_t>(std::floor(Y(local_pos) * wf)), 0l, width - 1);
  index_t k = std::clamp(
    static_cast<index_t>(std::floor(Z(local_pos) * wf)), 0l, width - 1);
  return data[l][i][j][k];
}

void
FMMTree::update(
  std::vector<MassSample>& samples,
  const std::function<Vec3(const Vec3&)>& field,
  const std::function<Mat3x3(const Vec3&)>& field_jacobian)
{
  using fmt::println;

  const auto& range = std::ranges::views::iota;

  index_t d = depth();
  index_t l = d - 1;
  index_t width = 1 << l;

  // Clear the tree leaf cells
  for (const index_t i : range(0, width)) {
    for (const index_t j : range(0, width)) {
      for (const index_t k : range(0, width)) {
        FMMCell& leaf = data[l][i][j][k];
        leaf.samples.clear();
        leaf.mass = 0.0;
        leaf.barycenter = { 0.0, 0.0, 0.0 };
      }
    }
  }

  // Populate the leaves and compute their masses and barycenter
  for (MassSample& sample : samples) {
    FMMCell& leaf = get_leaf_from_pos(sample.position);
    leaf.samples.emplace_back(sample);
    leaf.mass += sample.mass;
    leaf.barycenter += sample.mass * sample.position;
  }

  for (const index_t i : range(0, width)) {
    for (const index_t j : range(0, width)) {
      for (const index_t k : range(0, width)) {
        if (FMMCell& leaf = data[l][i][j][k]; !leaf.samples.empty())
          leaf.barycenter /= leaf.mass;
      }
    }
  }

  // Propagate mass and barycenter upward.
  // Unintuitively with the 7 nested for loops, this is done
  // in O(len(samples)) if depth is chosen to have as many
  // leaves as there are samples.
  for (const index_t ml : range(0, d - 1)) {
    l = d - 2 - ml;
    width = 1 << l;
    for (const index_t i : range(0, width)) {
      for (const index_t j : range(0, width)) {
        for (const index_t k : range(0, width)) {
          FMMCell& cell = data[l][i][j][k];
          cell.mass = 0.0;
          cell.barycenter = Vec3{ 0.0, 0.0, 0.0 };
          for (const index_t ci : { 2 * i, 2 * i + 1 }) {
            for (const index_t cj : { 2 * j, 2 * j + 1 }) {
              for (const index_t ck : { 2 * k, 2 * k + 1 }) {
                FMMCell& child = data[l + 1][ci][cj][ck];
                cell.mass += child.mass;
                cell.barycenter += child.mass * child.barycenter;
              }
            }
          }
          if (cell.mass > 0.0)
            cell.barycenter /= cell.mass;
        }
      }
    }
  }

  // Compute the field tensors. First, compute the contributions of the
  // interaction lists.
  for (l = 0; l < d; l++) {
    width = 1 << l;
    for (const index_t i : range(0, width)) {
      for (const index_t j : range(0, width)) {
        for (const index_t k : range(0, width)) {
          FMMCell& cell = data[l][i][j][k];
          if (cell.samples.empty())
            continue;
          cell.field_tensor.clear();
          for (auto& neighbor_ref : cell.interaction_list) {
            const FMMCell& neighbor = neighbor_ref.get();
            const Vec3 diff = cell.barycenter - neighbor.barycenter;
            const Vec3 field_intensity = field(diff);
            const Mat3x3 jacobian = field_jacobian(diff);
            cell.field_tensor.field += neighbor.mass * field_intensity;
            cell.field_tensor.jacobian += neighbor.mass * jacobian;
          }
        }
      }
    }
  }
  // Then, propagate the field tensors downward.
  for (l = 0; l < d - 1; l++) {
    width = 1 << l;
    for (const index_t i : range(0, width)) {
      for (const index_t j : range(0, width)) {
        for (const index_t k : range(0, width)) {
          const FMMCell& cell = data[l][i][j][k];
          for (const index_t ci : { 2 * i, 2 * i + 1 }) {
            for (const index_t cj : { 2 * j, 2 * j + 1 }) {
              for (const index_t ck : { 2 * k, 2 * k + 1 }) {
                FMMCell& child = data[l + 1][ci][cj][ck];
                child.field_tensor += cell.field_tensor;
              }
            }
          }
        }
      }
    }
  }
}

Array3D<FMMCell>&
FMMTree::operator[](index_t i)
{
  return data[i];
}
}

File: src/tree.py
from typing import Callable, List

import numpy as np

from .cell import FMMCell
from .sample import MassSample
from .utils import Mat3x3, Vec3


class FMMTree:
    def __init__(self, depth: int, size: float) -> None:
        """
        Initializes the FMMTree with the given depth and size.
        Constructs the tree structure and initializes the direct_neighbors and interaction_list fields.
        These are the only fields that are initialized, as they do not depend on the samples but only on the structure.

        Args:
            depth: depth of the tree
            size: size of the simulation cube (the cells' centroids are placed so that the cube is centered at the origin)

        Time complexity:
            O(8^depth) given the depth of the tree, which becomes O(len(samples)) if depth = O(log8(len(samples)).
        """
        # Initialize the data
        self.data = [
            [
                [
                    [
                        FMMCell(
                            np.array(
                                [
                                    size * ((i + 0.5) / 2**l - 0.5),
                                    size * ((j + 0.5) / 2**l - 0.5),
                                    size * ((k + 0.5) / 2**l - 0.5),
                                ]
                            ),
                            size / 2**l,
                        )
                        for k in range(2**l)
                    ]
                    for j in range(2**l)
                ]
                for i in range(2**l)
            ]
            for l in range(depth)
        ]

        # Construct direct_neighbors and interaction_list
        for l in range(depth):
            width = 2**l
            for i in range(width):
                for j in range(width):
                    for k in range(width):
                        cell = self[l][i][j][k]
                        for ni in [i - 1, i, i + 1]:
                            for nj in [j - 1, j, j + 1]:
                                for nk in [k - 1, k, k + 1]:
                                    if (
                                        0 <= ni < width
                                        and 0 <= nj < width
                                        and 0 <= nk < width
                                    ):
                                        cell.direct_neighbors.append(
                                            self[l][ni][nj][nk]
                                        )
                        if l > 0:
                            # Parent indices
                            pi, pj, pk = i // 2, j // 2, k // 2
                            for ni in range(max(0, 2 * pi - 2), min(width, 2 * pi + 4)):
                                for nj in range(
                                    max(0, 2 * pj - 2), min(width, 2 * pj + 4)
                                ):
                                    for nk in range(
                                        max(0, 2 * pk - 2), min(width, 2 * pk + 4)
                                    ):
                                        near_child = self[l][ni][nj][nk]
                                        if (
                                            max(
                                                abs(ni - i),
                                                abs(nj - j),
                                                abs(nk - k),
                                            )
                                            > 1
                                        ):
                                            cell.interaction_list.append(near_child)

    def depth(self):
        return len(self.data)

    def tree_size(self) -> int:
        # There are 2^(3k) cells in the kth grid of the list
        # The total is sum(l = 0 -> depth - 1) of 8^k, i.e. (8^depth - 1) / 7
        return (8 ** self.depth() - 1) // 7

    def space_size(self) -> float:
        return self[0][0][0][0].size

    def __getitem__(self, i: int) -> List[List[List[FMMCell]]]:
        return self.data[i]

    def __setitem__(self, i: int, value: List[List[List[FMMCell]]]):
        self.data[i] = value

    def pretty_print(self) -> None:
        def aux(i: int, j: int, k: int, l: int) -> None:
            if l == self.depth():
                print("  " * l + str(self[i][j][k]))
            else:
                print(" " * l + "FMMTree{")
                i2 = 2 * i
                j2 = 2 * j
                k2 = 2 * k
                for incr1 in [0, 1]:
                    for incr2 in [0, 1]:
                        for incr3 in [0, 1]:
                            aux(i2 + incr1, j2 + incr2, k2 + incr3, l + 1)
                print(" " * l + "}")

        aux(0, 0, 0, 0)

    def get_leaf_from_pos(self, pos: Vec3) -> FMMCell:
        """
        Returns the leaf cell containing the given position.
        """
        l = self.depth() - 1
        size = self.space_size()
        i = int((pos[0] / size + 0.5) * 2**l)
        j = int((pos[1] / size + 0.5) * 2**l)
        k = int((pos[2] / size + 0.5) * 2**l)
        return self[l][min(len(self[l]) - 1, max(0, i))][
            min(len(self[l]) - 1, max(0, j))
        ][min(len(self[l]) - 1, max(0, k))]

    def update(
        self,
        samples: List[MassSample],
        field: Callable[[Vec3], Vec3],
        field_jacobian: Callable[[Vec3], Mat3x3],
    ) -> None:
        """
        Updates the tree's content with the given samples and field function.
        The structure itself does not change, but the mass, barycenter, field tensors
        and distribution of samples in the leaves are updated.

        Args:
            samples: list of mass samples
            field: function of the distance between two points representing the field between them per unit mass
            field_jacobian: function of the distance between two points representing the jacobian of the field between them per unit mass

        Returns:
            None

        Time complexity:
            O(8^depth) given the depth of the tree, which becomes O(len(samples)) if depth = O(log8(len(samples)).
        """

        # Clear the tree leaf cells
        for i in range(2 ** (self.depth() - 1)):
            for j in range(2 ** (self.depth() - 1)):
                for k in range(2 ** (self.depth() - 1)):
                    leaf = self[-1][i][j][k]
                    if leaf.samples is not None:
                        del leaf.samples[:]
                    else:
                        leaf.samples = []
                    leaf.mass = 0
                    leaf.barycenter = np.zeros(3)

        # Populate the leaves and compute their masses and barycenter
        for sample in samples:
            leaf = self.get_leaf_from_pos(sample.pos)
            assert leaf.samples is not None
            leaf.samples.append(sample)
            leaf.mass += sample.mass
            leaf.barycenter += sample.mass * sample.pos

        for i in range(2 ** (self.depth() - 1)):
            for j in range(2 ** (self.depth() - 1)):
                for k in range(2 ** (self.depth() - 1)):
                    cell = self[self.depth() - 1][i][j][k]
                    if cell.samples is not None and len(cell.samples) != 0:
                        cell.barycenter /= cell.mass

        # Propagate mass and barycenter upward.
        # Unintuitively with the 7 nested for loops, this is done
        # in O(len(samples)) if depth is chosen to have as many
        # leaves as there are samples.
        for l in range(self.depth() - 2, -1, -1):
            for i in range(2**l):
                for j in range(2**l):
                    for k in range(2**l):
                        cell = self[l][i][j][k]
                        cell.mass = 0
                        cell.barycenter = np.zeros(3)
                        for ci in [2 * i, 2 * i + 1]:
                            for cj in [2 * j, 2 * j + 1]:
                                for ck in [2 * k, 2 * k + 1]:
                                    child = self[l + 1][ci][cj][ck]
                                    cell.mass += child.mass
                                    cell.barycenter += child.mass * child.barycenter
                        if cell.mass != 0:
                            cell.barycenter /= cell.mass

        # Compute the field tensors. First, compute the contributions of the interaction lists.
        for l in range(self.depth()):
            for i in range(2**l):
                for j in range(2**l):
                    for k in range(2**l):
                        cell = self[l][i][j][k]
                        if cell.samples is not None and len(cell.samples) == 0:
                            continue
                        cell.field_tensor.clear()
                        for neighbor in cell.interaction_list:
                            diff = cell.barycenter - neighbor.barycenter
                            field_intensity = field(diff)
                            jacobian = field_jacobian(diff)
                            cell.field_tensor.field += neighbor.mass * field_intensity
                            cell.field_tensor.jacobian += neighbor.mass * jacobian
        # Then, propagate the field tensors downward.
        for l in range(self.depth() - 1):
            for i in range(2**l):
                for j in range(2**l):
                    for k in range(2**l):
                        cell = self[l][i][j][k]
                        for ci in [2 * i, 2 * i + 1]:
                            for cj in [2 * j, 2 * j + 1]:
                                for ck in [2 * k, 2 * k + 1]:
                                    child = self[l + 1][ci][cj][ck]
                                    child.field_tensor += cell.field_tensor
                                    child.field_tensor.field += (
                                        cell.field_tensor.jacobian
                                        @ (child.barycenter - cell.barycenter)
                                    )
